source/wifi/wifi_hal.c

```diff
@@ -6700,10 +6722,175 @@ INT wifi_setRadioStatsEnable(INT radioIndex, BOOL enable)
+        fprintf(stderr, "failed to parse nested attributes!\n");
+        return NL_SKIP;
+    }
+
+    nla_for_each_nested(tidattr, sinfo[NL80211_STA_INFO_TID_STATS], rem)
+    {
+        stats_entry = &out->tid_array[tid_index];
+
+        stats_entry->tid = tid_index;
+        stats_entry->ac = _tid_ac_index_get[tid_index];
+
+        if(sinfo[NL80211_STA_INFO_TID_STATS])
+        {
+            if(nla_parse_nested(stats_info, NL80211_TID_STATS_MAX,tidattr, tid_policy)) {
+                printf("failed to parse nested stats attributes!");
+                return;
+            }
+        }
+
+        if(stats_info[NL80211_TID_STATS_TX_MSDU])
+            stats_entry->num_msdus = (unsigned long long)nla_get_u64(stats_info[NL80211_TID_STATS_TX_MSDU]);
+
+            tid_index++;
+            printf("all vals are %d , %d , %lld, \n",stats_entry->tid, stats_entry->ac , stats_entry->num_msdus);
+    }
+    //ToDo: sum_time_ms, ewma_time_ms
+
+    return NL_SKIP;
+}
+#endif
+
 INT wifi_getApAssociatedDeviceTidStatsResult(INT radioIndex,  mac_address_t *clientMacAddress, wifi_associated_dev_tid_stats_t *tid_stats,  ULLONG *handle)
 {
-    // TODO Implement me!
-    return RETURN_ERR;
+#ifdef HAL_NETLINK_IMPL
+   Netlink nl;
+   char  if_name[10];
+   char mac_addr[MAC_ALEN];
+
+   snprintf(if_name,sizeof(if_name),"wlan%d",radioIndex);
+
+   nl.id = initNl80211(&nl);
+
+
+   if (nl.id < 0) {
+       fprintf(stderr, "Error initializing netlink \n");
+       return -1;
+   }
+
+   struct nl_msg* msg = nlmsg_alloc();
+
+   if (!msg) {
+       fprintf(stderr, "Failed to allocate netlink message.\n");
+       nlfree(&nl);
+       return -2;
+   }
+
+   genlmsg_put(msg,
+              NL_AUTO_PORT,
+              NL_AUTO_SEQ,
+              nl.id,
+              0,
+              0,
+              NL80211_CMD_GET_STATION,
+              0);
+
+   if(mac_addr_aton(mac_addr, clientMacAddress)) {
+       printf("invalid mac address\n");
+       return 0;
+   }
+
+   nla_put(msg, NL80211_ATTR_MAC, MAC_ALEN, mac_addr);
+   nla_put_u32(msg, NL80211_ATTR_IFINDEX, if_nametoindex(if_name));
+   nl_cb_set(nl.cb,NL_CB_VALID,NL_CB_CUSTOM,tidStats_callback,tid_stats);
+   nl_send_auto(nl.socket, msg);
+   nl_recvmsgs(nl.socket, nl.cb);
+   nlmsg_free(msg);
+   nl_cb_put(nl.cb);
+   nlfree(&nl);
+   return RETURN_OK;
+#else
+//iw implementation
+#define TID_STATS_FILE "/tmp/tid_stats_file.txt"
+#define TOTAL_MAX_LINES 50
+
+    char buf[256] = {'\0'}; /* or other suitable maximum line size */
+    char if_name[10]= "wlan0";
+    FILE *fp=NULL;
+    char pipeCmd[1024]= {'\0'};
+    int lines,tid_index=0;
+    char mac_addr[20] = {'\0'};
+
+    wifi_associated_dev_tid_entry_t *stats_entry;
+
+    snprintf(if_name,sizeof(if_name),"wlan%d",radioIndex);
+
+    strcpy(mac_addr,clientMacAddress);
+
+    snprintf(pipeCmd,sizeof(pipeCmd),"iw dev %s station dump -v > "TID_STATS_FILE,if_name);
+    fp= popen(pipeCmd,"r");
+
+    if(fp == NULL)
```

Comment:
I would like to have this line 'perror("popen failed");' for all popen error condition

Reply:
Noted